keps/sig-auth/4193-tokenrequest-requester-info-enhancements/README.md

```diff
@@ -0,0 +1,616 @@
+TokenRequest API, in a similar manner to how the ServiceAccount, Pod or Secret is referenced.
+
+Additionally, to provide a robust means of tracking token usage within the audit log we can embed a unique identifier for
+each token which is can then also be recorded in future audit entries made by this token.
+
+## Motivation
+
+### Goals
+
+* Embedding information about the Node that a pod is running on into signed JWTs.
+* Make it easier to track the actions a single token has taken, and cross-reference that back to the origin of the token
+  (via audit log inspection).
+* Provide a means of checking whether a Pod's token is associated with the same Node as it was associated with when the
+  initial TokenRequest was made (via an extra field that can be observed from the TokenReview API).
+
+### Non-Goals
+
+* Embedding requester information. This is discussed further in the alternatives
+  considered section, and a future KEP may revisit this.
+* Embedding information beyond the immutable Node name and UID into the token. We aim to mimic what is done with the ref fields
+  for secret, pod and serviceaccount (not introduce any additional properties).
+* Changing default behaviour of the SA authenticator to enforce the referenced Node object still exists.
+
+## Proposal
+
+### Embedding Pod's bound Node information in tokens
+
+The kube-apiserver will be extended to automatically embed the `name` and `uid` of the *Node* a Pod is associated
+with (via `spec.nodeName`) in generated tokens when a TokenRequest `create` call is serviced.
+
+As the 'pod' is already available in this area of code, which contains the `nodeName`, we will just need to plumb
+through a Getter for Node objects into the TokenRequest storage layer so the node's UID can be fetched, similar to
+what is done for pod & secret objects.
+
+### Allowing ServiceAccount tokens to be bound to a Node object
+
+Similar to how a token can be bound to a Pod or Secret object, we will also extend the TokenRequest API to allow
+binding directly to Node objects (without needing to bind to a Pod as well).
+
+This allows users to obtain a token that is tied specifically to the *Node* objects lifecycle, i.e. when the Node
+object is deleted, the token will be invalidated.
+
+### Extending TokenReview to allow cross-checking the embedded Node information with existing Node objects
+
+The SA authenticator will also be extended to check whether a JWT's embedded node information is still valid/current,
+and if not, will reject the review/indicate to the client that the token mismatches with the current state of Nodes.
+
+For tokens bound to Pod objects, to avoid unexpected breaking changes in behaviour, this will not only be protected by
+the feature flag whilst the feature is graduating to GA, but will also be gated behind an additional kube-apiserver flag
+`--service-account-validate-node-info` which defaults to 'off'. This avoids changing the previous behaviour for
```

Comment:
This is only a concern until all supported kube-apiservers know about node bindings, right?  After that, an old server will properly reject when nodes are gone and new servers will always have to accept tokens without node bindings.